ens-improvement-proposals/ensip-15-normalization-standard.md

```diff
@@ -0,0 +1,343 @@
+# ENSIP-15: ENS Name Normalization Standard
+
+| **Author**  | Andrew Raffensperger \<raffy@me.com> |
+| ----------- | ------------------------------------ |
+| **Status**  | Draft                                |
+| **Created** | 2023-04-03                           |
+
+## Abstract
+
+This ENSIP standardizes Ethereum Name Service (ENS) name normalization process outlined in [ENSIP-1 § Name Syntax](./ensip-1-ens#name-syntax).
+
+## Motivation
+
+* Since [ENSIP-1](./ensip-1-ens) (originally [EIP-137](https://eips.ethereum.org/EIPS/eip-137)) was finalized in 2016, Unicode has [evolved](https://unicode.org/history/publicationdates.html) from version 8.0.0 to 15.0.0 and incorporated many new characters, including complex emoji sequences. 
+* ENSIP-1 does not state the version of Unicode.
+* ENSIP-1 implies but does not state an explicit flavor of IDNA processing. 
+* [UTS-46](https://unicode.org/reports/tr46/) is insufficient to normalize emoji sequences. Correct emoji processing is only possible with [UTS-51](https://www.unicode.org/reports/tr51/).
+* Validation tests are needed to ensure implementation compliance.
+* The success of ENS has encouraged spoofing via the following techniques:
+	1. Insertion of zero-width characters.
+	1. Using names which normalize differently between algorithms. 
+	1. Using names which appear differently between applications and devices.
+	1. Substitution of confusable (look-alike) characters.
+	1. Mixing incompatible scripts.
+
+## Specification
+
+### Versioning
+* Unicode version `15.0.0`
+* [spec.json](./ensip-15/spec.json) contains all necessary data for normalization.
+* [nf.json](./ensip-15/nf.json) contains all necessary data for [Unicode Normalization Forms](https://unicode.org/reports/tr15/) NFC and NFD.
+
+### Algorithm
+
+* Normalization is the process of canonicalizing a name before for hashing.
+* It is idempotent: applying normalization multiple times produces the same result.
+* For user convenience, leading and trailing whitespace should be trimmed before normalization, as all whitespace codepoints are disallowed.  Inner characters should remain unmodified.
+* No string transformations (like case-folding) should be applied.
+
+1. Split the name into [labels](./ensip-1-ens#name-syntax).

```

Comment:
Sorry for the late edits — okay, adding Split and Join sections

I had left them undefined since it's the same mechanics as ENSIP-1 (which I referenced.)

Reply:
There are Join and Split sections:

* Split
    * Partition a name into labels, separated by `2D (.) FULL STOP`, and return the resulting array.
    * Example: `"abc.123.eth"` → `["abc", "123", "eth"]`
* Join
    * Assemble an array of labels into a name, inserting `2D (.) FULL STOP` between each label, and return the resulting string.
    * Example: `["abc", "123", "eth"]` → `"abc.123.eth"`